enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+
+- As an OCP engineer, I want the NHC to consume the same logic for detecting
+ failures, to avoid code duplication and provide a consistent experience
+  compared to MHC.
+
+- As an OCP engineer, I want the NHC to use the same interface for initiating
+ recovery, so that we can reuse existing mechanisms (like [poison pill] and metal3) and
+  avoid code duplication.
+
+To fulfil the above we create a [Node Health Check custom resource (NCR)](anchor todo) that specifies:
+1. health criteria: list of criterias and selector
+2. template: a template of an [external custom resource (remediation CR)](link todo)
+ to create in response to a match.
+
+### Unhealthy criteria
+A node target is unhealthy when the node meets the unhealthy node conditions criteria defined.
+If any of those criterias are met for longer than given timeouts, the controller
+creates a remediation CR according to the template in the NHC CRD.
+Timeouts will be defined as part of the NHC CRD .
+
+### Implementation Details
+
+#### NodeHealthCheck CRD
+- Enable watching a node based on a label selector. 
+- Enable defining an unhealthy node criteria (based on a list of node conditions).
+- Enable setting a threshold of unhealthy nodes. If the current number is at or above this threshold no further action will take place. This can be expressed as an int or as a percentage of the total targets in the pool.
+
+E.g:
+- Create a remedy CR when node is `ready=false` or `ready=Unknown` condition for more than 5m.
+- I want to avoid triggering remediation if 40% or more of the targets of this pool are unhealthy at the same time.
+
+
+```yaml
+apiVersion: remediation.medik8s.io/v1alpha1
+kind: NodeHealthCheck
+metadata:
+  name: example
+  # this CRD is cluster-scope, namespace is not needed  
+spec:
+  maxUnhealthy: "49%"
+  selector:
+    matchExpressions:
+      key: node-role.kubernetes.io/worker
+      operator: exists
+  unhealthyConditions:
+    - type: "Ready"
+      status:  "Unknown"
+      duration: "300s"
+   - type: "Ready"
+     status:  "False"
+     duration: "300s"
+  remediationTemplate:
+    kind: PoisonPillRemediationTemplate
+    apiVersion: poison-pill.medik8s.io/v1alphaX
+    name: M3_REMEDIATION_GROUP
+status:
+  healthyNodes: 5
+  obeservedNodes: 5
+```
+
+#### NodeHealthCheck controller
+Watch:
+- Watch NodeHealthCheck resources
+- Watch nodes with an event handler e.g controller runtime `EnqueueRequestsFromMapFunc` which returns NodeHealthCheck resources.
+
+Reconcile:
+- Periodically enqueue a reconcile request
+- Fetch all nodes targets. E.g:
+
+
+- Calculate the number of unhealthy targets.
+- Compare current number against `maxUnhealthy` threshold and temporary short
+ circuits if the threshold is met.
+- Create remedy objects for unhealthy targets according to the template in NHC object
+
+### Notes/Constraints
+
+The functionality of this controller have a lot in common with the Machine Health
+Check controller except that its creating objects and not
```

Comment:
We did approach CAPI with that idea, however there was no enthusiasm at the time.
This is not hugely surprising since it effectively deemphasises the role of Machines and MAPI.

We're still open to the idea of adapting the existing controller (that's basically how NHC was created), however I would have concerns about the speed at which progress can be made.  MHC was a fine MVP 2.5 years ago but it really hasn't been allowed to progress much since.

Reply:
Let's capture this info in the doc.

Reply:
I don't think this has been captured yet